lm_eval/models/mlx_llms.py

```diff
@@ -0,0 +1,286 @@
+                )
+
+            # Pad to the nearest multiple of 8 or the maximum length
+            pad_to = 8
+            max_length_in_batch = pad_to * ((max(lengths) + pad_to - 1) // pad_to)
+            max_length_in_batch = min(max_length_in_batch, self.max_tokens)
+
+            batch_arr = np.zeros((current_batch_size, max_length_in_batch), np.int32)
+
+            for j in range(current_batch_size):
+                truncated_length = min(lengths[j], self.max_tokens)
+                batch_arr[j, :truncated_length] = full_sequences[j][:truncated_length]
+                lengths[j] = (
+                    truncated_length  # Update lengths to match truncated lengths
+                )
+            batch = mx.array(batch_arr)
+            non_padding_lengths = mx.array(lengths)
+            prompt_lengths = mx.array(prompt_lengths)
+
+            shifted_padded_full_sequence = batch[
+                :, :-1
+            ]  # all but the last token for each sequence
+            logits = self.model(shifted_padded_full_sequence)
+            logits = logits.astype(mx.float32)
+
+            # log softmax probabilities
+            log_probs = mx.log(mx.softmax(logits, axis=-1))
+
+            # Create mask to exclude padding and inputs
+            mask_width = shifted_padded_full_sequence.shape[1]
+            flattened_token_indices = mx.arange(mask_width)
+            token_indices = flattened_token_indices[None, :]
+            mask = mx.logical_and(
+                token_indices >= prompt_lengths[:, None],
+                token_indices < non_padding_lengths[:, None],
+            )
+
+            batch_greedy_tokens = logits.argmax(axis=-1)
+            # A sequence of 1s or 0's the same width as the batch, where 1 indicates the target token is the same
+            # as the greedily-generated token (determined efficiently via argmax on token probabilities)
+            masked_indicator_values = (
+                batch_greedy_tokens == shifted_padded_full_sequence
+            ) * mask
+
+            # A sequence of booleans indicating whether the sum of indicator values is equal to corresponding
+            # target length
+            batch_target_is_greedy_values = masked_indicator_values.sum(
+                axis=-1
+            ) == mx.array(non_padding_lengths)
+
+            # Iterate over question, answer pairs, their log softmax logits, and their greedy values
+            for idx, (is_greedy, log_prob) in enumerate(
+                zip(batch_target_is_greedy_values, log_probs)
+            ):
+                prompt_length = prompt_lengths[idx].item()
+                full_sequence_length = non_padding_lengths[idx].item()
+                target_length = full_sequence_length - prompt_length
+                context = context_batch[idx]
+                continuation = continuation_batch[idx]
+
+                # Extract target sequence and log prob scores that correspond to the predicted probability for it
```

Comment:
Right now we are extracting the log probs of `<continuation><eos>` while we only need the log probs corresponding with `<continuation>`

Reply:
I was using this mainly to ensure the answers and their scores were being returned in the original order of the function being called.  Is this still relevant if this is all moved to _loglikelihood_tokens?

Reply:
I think you are in including the <eos> token in the target logprobs?:
https://github.com/EleutherAI/lm-evaluation-harness/blob/cc48c102251514af418458876da7b230ac3140d0/lm_eval/models/mlx_llms.py#L213C1-L221